students/kristianjf/lesson09/mailroom_5.py

```diff
@@ -12,44 +12,68 @@
 '''
-# self.DATA_STRUCTURE = [('Jeff Bezos', [200, 500, 1000]), \
-#                     ('Bill Gates', [1000]), \
-#                     ('Steve Jobs', [20, 50, 100])]
-# self.DATA_STRUCTURE_DICTS = [{'name': name, 'donation': donation} for name, donation in self.DATA_STRUCTURE]
-# # self.DATA_STRUCTURE_DICT = [ \
-# #                         {'name': 'Jeff Bezos', 'donation': [200, 500, 1000]},\
-# #                         {'name': 'Bill Gates', 'donation': [1000]},\
-# #                         {'name': 'Steve Jobs', 'donation': [20, 50, 100]}\
-# #                         ]
-
 class Donor:
     def __init__(self, name, donations):
         self._name = name
-        self._donations = [] if donations is None else donations
+        self._normalized_name = ''
+        self._donations = donations if isinstance(donations, list) else [donations]
+    def __lt__(self, other):
+        return sum(self.donations) < sum(other.donations)
     @property
     def name(self):
         return self._name
     @property
     def donations(self):
         return self._donations
+    @property
+    def normalized_name(self):
+        'Validate and Return Name Input'
+        full_name_cap = ''
+        try:
+            for name in self.name.split():
+                assert name.isalpha()
+                full_name_cap += f'{name.capitalize()} '
+            full_name = full_name_cap.strip()
+        except AssertionError:
+            print('Invalid Name. Found non-alphabetic characters')
+        else:
+            return full_name
+        return self._normalized_name
     def donate(self, new_donation):
         self.donations.append(new_donation)
     def metrics(self):
-        self.num_d = len(self.donations)
-        self.total_d = sum(self.donations)
-        self.avg_d = self.total_d / self.num_d
+        num_d = len(self.donations)
+        total_d = sum(self.donations)
+        avg_d = total_d / num_d
+        return num_d, total_d, avg_d
+
+    def send_letter(self):
+        '''
+        Try to use a dict and the .format() method to do the letter as one big \
+        template rather than building up a big string in parts.
+        In this version, add a function (and a menu item to invoke it), that goes \
+        through all the donors in your donor data structure, generates a thank you \
+        letter, and writes it to disk as a text file.
+        '''
+        filename = '_'.join(self.name.split())+'.txt'
+        with open(filename, 'w') as outfile:
+            outfile.write(self.output_string().format(name=self.name, donation=self.donations))
+            print(f'Writing: {filename}')
+
+    def output_string(self):
+        'Output string for letters'
+        format_string = 'Dear {name},\n\nThank you for your generous donation of ' \
+                        '${donation}. Please send us more money at your earliest ' \
+                        'convenience.'
+        return format_string
```

Comment:
This return value looks like a fixed string, so I'm not sure if this needs to be a method? Could have been just a class attribute?